torch/_inductor/scheduler.py

```diff
@@ -178,6 +178,29 @@ def can_free(self):
                 return False
         return True
 
+    def max_alive_variables(self):
+        graph = self._body.root_block.graph
+        die_now = []
+        dead_variables = set()
+        for n in reversed(graph.nodes):
+            # Use n.name, we don't asume that DCE has run before this function
+            used_vars = set(itertools.chain((node.name for node in n.args if isinstance(node, torch.fx.Node)),
+                                            (node.name for node in n.kwargs.values() if isinstance(node, torch.fx.Node)),
+                                            (n.name,)))
+            die_now.append(used_vars - dead_variables)
+            dead_variables |= die_now[-1]
+
+        # We don't want to track the ops variable
+        die_now[-1].add("ops")
+        max_alive = 0
+        alive = set()
+        for n, die in zip(graph.nodes, reversed(die_now)):
+            var = n.name
+            alive.add(var)
+            alive -= die
+            max_alive = max(max_alive, len(alive))
+        return max_alive
```

Comment:
I believe this can be done in a single pass:
```python
        max_alive = 0
        live_variables = set()
        for n in reversed(graph.nodes):
            live_variables.discard(n.name)
            used_vars = set(itertools.chain((node.name for node in n.args if isinstance(node, torch.fx.Node)),
                                            (node.name for node in n.kwargs.values() if isinstance(node, torch.fx.Node))))
           
            live_variables += used_vars
            max_live = max(max_live, len(live_variables))

        return max_alive
```

The same principle can probably be applied to `estimate_reg_usage`

Reply:
I don't think I quite understand that approach. Variables die whenever they are last used, and that often happens when they are used as an argument to a function. In order to figure out when are they last used, you need to have a look at the operations in reverese. Your algorithm says that every fresh new variable is dead and you discard it?
Can you elaborate?

Reply:
> Your algorithm says that every fresh new variable is dead and you discard it?

Since we're iterating in reverse order, a new variable means a variable that does not exist from that node onwards. It is equivalent to `alive.add(var)` but inverted.

The concept here is that variables are alive from the point they are first seen (which in reverse is their last usage) and the point they are assigned, before which they do not exist.

Reply:
I completely missed that you were iterating the graph in reverse order lol
You are completely right. Your version of the algorithm is much simpler! Let me implement that one.